heft-plugins/heft-webpack4-plugin/src/WebpackPlugin.ts

```diff
@@ -1,157 +1,186 @@
 const WEBPACK_DEV_SERVER_ENV_VAR_NAME: string = 'WEBPACK_DEV_SERVER';
 
-interface IWebpackVersions {
-  webpackVersion: string;
-  webpackDevServerVersion: string;
-}
-
 /**
  * @internal
  */
-export class WebpackPlugin implements IHeftPlugin {
-  public readonly pluginName: string = PLUGIN_NAME;
+export default class WebpackPlugin implements IHeftTaskPlugin {
+  private _webpack: typeof TWebpack | undefined;
+  private _webpackVersions: IWebpackVersions | undefined;
+  private _loadedWebpackConfiguration: IWebpackConfiguration | null | undefined;
+
+  public readonly accessor: IWebpackPluginAccessor = {
+    onEmitWebpackVersionsHook: new AsyncParallelHook(['webpackVersions']),
+    onConfigureWebpackHook: new AsyncSeriesWaterfallHook(['webpackConfiguration']),
+    onAfterConfigureWebpackHook: new AsyncParallelHook(['webpackConfiguration']),
+    onEmitStatsHook: new AsyncParallelHook(['webpackStats'])
+  };
+
+  public apply(taskSession: IHeftTaskSession, heftConfiguration: HeftConfiguration): void {
+    // These get set in the run hook and used in the onConfigureWebpackHook
+    let production: boolean;
+    let serveMode: boolean;
+    let watchMode: boolean;
+
+    this.accessor.onConfigureWebpackHook!.tapPromise(
+      PLUGIN_NAME,
+      async (existingConfiguration: IWebpackConfiguration | null) => {
+        if (existingConfiguration) {
+          taskSession.logger.terminal.writeVerboseLine(
+            'Skipping loading webpack config file because the webpack config has already been set.'
+          );
+          return existingConfiguration;
+        } else {
+          const configurationLoader: WebpackConfigurationLoader = new WebpackConfigurationLoader(
+            taskSession.logger,
+            production,
+            serveMode
+          );
+          return await configurationLoader.tryLoadWebpackConfigAsync(heftConfiguration.buildFolder);
+        }
+      }
+    );
+
+    taskSession.hooks.clean.tapPromise(PLUGIN_NAME, async (cleanOptions: IHeftTaskCleanHookOptions) => {
+      // Obtain the finalized webpack configuration
+      const webpackConfiguration: IWebpackConfiguration | null = await this._getWebpackConfigurationAsync();
+      if (webpackConfiguration) {
+        const webpackConfigurationArray: IWebpackConfigurationWithDevServer[] = Array.isArray(
+          webpackConfiguration
+        )
+          ? webpackConfiguration
+          : [webpackConfiguration];
+
+        // Add each output path to the clean list
+        for (const config of webpackConfigurationArray) {
+          if (config.output?.path) {
+            cleanOptions.addDeleteOperations({
+              sourceFolder: config.output.path
+            });
+          }
+        }
+      }
+    });
+
+    taskSession.hooks.run.tapPromise(PLUGIN_NAME, async (runOptions: IHeftTaskRunHookOptions) => {
+      production = runOptions.production;
+      // TODO: Support watch mode
+      watchMode = false;
+      // TODO: Support serve mode
```

Comment:
These are already implemented.

Reply:
Hmm... may need to rejig how we provide the "serve"/"watch"/"production" parameters. Currently they would come in the hook options, but that doesn't work when you need them in the external plugin hooks.

Reply:
Implemented, except that the watch mode is non-compliant with the surrounding Heft infrastructure, and there isn't a way to pass it from Heft right now.

Reply:
Yes. Right now there isn't really a full story around how to run the watch/serve mode stuff just yet. I will need to come back and address these when I take a closer look at that.